t/app/controller/admin_permissions.t

```diff
@@ -0,0 +1,226 @@
+
+        $mech->submit_form_ok( { with_fields => {
+            name => 'Test Updated User 2',
+            email => $user2->email,
+            body => $user2->from_body->id,
+            phone => '',
+            flagged => undef,
+        } } );
+        $user2->discard_changes;
+        is $user2->name, 'Test Updated User 2', 'name changed';
+    };
+
+    $oxfordshireuser->user_body_permissions->create({
+        body => $oxfordshire,
+        permission_type => 'user_manage_permissions',
+    });
+
+    subtest "Users can edit permissions" => sub {
+        is $user2->user_body_permissions->count, 0, 'user2 has no permissions';
+
+        $mech->get_ok("/admin/user_edit/$user2_id");
+        $mech->content_contains('Moderate report details');
+
+        $mech->submit_form_ok( { with_fields => {
+            name => $user2->name,
+            email => $user2->email,
+            body => $user2->from_body->id,
+            phone => '',
+            flagged => undef,
+            "permissions[moderate]" => 'on',
+            "permissions[report_edit_category]" => undef,
+            "permissions[report_edit_priority]" => undef,
+            "permissions[report_inspect]" => undef,
+            "permissions[report_instruct]" => undef,
+            "permissions[contribute_as_another_user]" => undef,
+            "permissions[contribute_as_body]" => undef,
+            "permissions[user_edit]" => undef,
+            "permissions[user_manage_permissions]" => undef,
+            "permissions[user_assign_areas]" => undef,
+        } } );
+
+        ok $user2->has_permission_to("moderate", $user2->from_body->id), "user2 has been granted moderate permission";
+    };
+
+    subtest "Unsetting user from_body removes all permissions " => sub {
+        is $user2->user_body_permissions->count, 1, 'user2 has 1 permission';
+
+        $mech->get_ok("/admin/user_edit/$user2_id");
+        $mech->content_contains('Moderate report details');
+
+        $mech->submit_form_ok( { with_fields => {
+            name => $user2->name,
+            email => $user2->email,
+            body => undef,
+            phone => '',
+            flagged => undef,
+            "permissions[moderate]" => 'on', # NB tick box is left on deliberately
+            "permissions[report_edit_category]" => undef,
+            "permissions[report_edit_priority]" => undef,
+            "permissions[report_inspect]" => undef,
+            "permissions[report_instruct]" => undef,
+            "permissions[contribute_as_another_user]" => undef,
+            "permissions[contribute_as_body]" => undef,
+            "permissions[user_edit]" => undef,
+            "permissions[user_manage_permissions]" => undef,
+            "permissions[user_assign_areas]" => undef,
+        } } );
+
+        is $user2->user_body_permissions->count, 0, 'user2 has had permissions removed';
+    };
+};
+
+$mech->log_out_ok;
+
+$mech->delete_user( $user );
```

Comment:
If you put these lines in an `END`, they should run even if a test in this file causes it to die.